common/unsafe/src/main/java/org/apache/spark/unsafe/memory/OnHeapMemoryBlock.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.unsafe.memory;
+
+import org.apache.spark.unsafe.Platform;
+
+/**
+ * A consecutive block of memory with a long array on Java heap.
+ */
+public final class OnHeapMemoryBlock extends MemoryBlock {
+
+  private final long[] array;
+
+  public OnHeapMemoryBlock(long[] obj, long offset, long size) {
+    super(obj, offset, size);
+    this.array = obj;
+    assert(offset - Platform.LONG_ARRAY_OFFSET + size <= obj.length * 8L) :
```

Comment:
So, ```obj.length * 8L``` should also change to ```size```, right?

Reply:
This condition is correct, but the message is not correct. Good catch.

Reply:
Hi, @kiszk. I still doubt about this ```assert```'s correctness. Below is my understanding, please feel free to point my wrong spot out. 

This method ```public OnHeapMemoryBlock(long[] obj, long offset, long size)``` could be called from three paths:
1. called from ```public OnHeapMemoryBlock(long size)```
In this case, this ```assert``` certainly to be true, and ```offset``` is equal to ```Platform.LONG_ARRAY_OFFSET```.

2. called directly from other methods 
In this case, I think the caller will mostly set ```offset``` to be ```Platform.LONG_ARRAY_OFFSET```. And the ```assert``` is also right, but the result due to caller's settings.

3. called from ``` public MemoryBlock subBlock(long offset, long size)```
In this case, a sub block can only use the memory which a super block could use, which is super block's ```size```, rather than ```obj.length * 8L``` (as we reach agreement before).
So, I think, here, the ```assert``` should be ```(offset - Platform.LONG_ARRAY_OFFSET + sub block's size <= super block's size)```. More precisely, we should ```assert(offset >=  Platform.LONG_ARRAY_OFFSET)```.

Reply:
Thank you for your explanation. Now, we focus on the path 3. I understand what you pointed out. How the following code should work?
```
val mb = new OnHeapMemoryBlock(new long[16]), Platform.LONG_ARRAY_OFFSET, 8);
val mb1 = mb.subBlock(0, 96);
```

According to the name `subBlock()`, I understand that you would like to ensure `sub block's size <= super block's size` (`not (96 <= 8)`)` logically.
I would like to allow us to the above code since the array actually reserved enough space (`96 <= 128`). This is because I have already seen such a code when we would use `MemoryBlock` for `

I. Add additional method `block()` (or different name) with this assertion. Keep `subBlock()` with a new assert that you proposed.
2. Use another name `block()` (or different name) instead of `subBlock()` with this assertion.

WDYT? cc: @cloud-fan